### PR TITLE
Update auth and prediction routes

### PR DESCRIPTION
**Pull request by collinschreyer-dev**
## **Description of Changes**

This pull request introduces functionality for filtering solicitations by office based on the user's email domain. Specifically, the updates include:

1. **Mapping Email Domains to Offices**:
   - We created a mapping between email domains and specific offices using the function `getOfficeFromEmail`. This dynamically determines the user's office based on their email address.
   - The mappings ensure consistent terminology and handle variations in office names.

2. **Enhancements in Filtering Logic**:
   - The filtering logic now applies the computed `office` from the user's email dynamically during the query.
   - Additional safeguards were introduced to ensure that filters are normalized and consistently applied.

3. **Dynamic Parent Agency Resolution**:
   - A helper function `getParentAgencyFromOffice` was added to compute the parent agency dynamically from the office name.

4. **User Role and Permissions Handling**:
   - For non-admin users, filtering logic ensures that solicitations are scoped to the user's agency and office, with the ability to override these filters via request parameters.

---

## **New Helper Functions**

1. **`getOfficeFromEmail`**:
   - Determines the user's office based on their email domain.
2. **`getParentAgencyFromOffice`**:
   - Maps the user's office to its parent agency.

---

## **Data Structures Added**

1. **`topLevelAgencyMap`**:
   - Maps top-level agencies to their sub-agencies.
2. **`agencyNameVariations`**:
   - Handles variations of agency names to ensure consistent mapping.

---

## **Updated Functions for User Management**

1. **`updateMAXUser`, `updateUser`, `createUser`**:
   - These functions now include logic to extract and update the office dynamically based on the user's email.

---

## **Enhanced Logging**

- Added detailed logs in the following functions:
  - **`createMAXUser`** and **`casStage2`**:
    - Now include office and agency information in the logs for better traceability.

---

## **Prediction Filtering**

1. **`getPredictions`**:
   - Includes dynamic office-based filtering, ensuring users see relevant predictions based on their assigned office.
2. **`prediction.routes.js`**:
   - Updated to support additional filters such as `userEmail`.

---

## **Verification Process**

We thoroughly verified this implementation by:

1. **Testing with All Email Variations**:
   - Logged into the development environment using emails with all possible domain variations for mapped offices.
   - Verified that the computed office name matched the expected result for each email domain.

2. **Database Cross-Check**:
   - Queried the database directly to count the number of solicitations for each office.
   - Compared the database results with the API responses to confirm that the correct number of solicitations appeared for each office filter.

3. **Screenshots**:
   - Screenshots were taken during the verification process to document how the filters were applied and confirm the expected results.

---

## **Screenshots**

Screenshot 2024-11-27 at 8 09 23 AM

Screenshot 2024-11-27 at 8 09 34 AM

---

## **Notes**

- The logic adheres to consistent naming conventions for offices and ensures that solicitation results are scoped appropriately based on user roles.
- These changes are necessary to support filtering users and predictions by office, ensuring accurate categorization and access.

### CODE REVIEW

File: server/routes/auth.routes.js [hunk 1]
```diff
@@ -119,14 +134,15 @@ function getGovernmentEmail(emails) {
...
-  //console.log("Login.gov user:", loginGovUser)
-
-  const gov_email = getGovernmentEmail(loginGovUser.all_emails || [])
+  logger.log('info', 'Creating new Login.gov user', {
+    email: loginGovUser.email,
+    office: getOfficeFromEmail(loginGovUser.email),
+    tag: 'createUser'
+  })
 
-  const user_email = gov_email || loginGovUser.email
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:129]: Why did you remove the user_email? It gets placed in the user_data.

File: server/routes/auth.routes.js [hunk 2]
```diff
@@ -135,20 +151,107 @@ function createUser(loginGovUser) {
...
+  // Treasury Mappings
+  if (emailLower.includes('irs.gov')) return 'INTERNAL REVENUE SERVICE';
+  if (emailLower.includes('usmint.treas.gov')) return 'US MINT';
+
+
+  return null;
+ }
+
+
+function getParentAgencyFromOffice(office) {
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:251]: I don't understand this path, when we have a function "grabAgencyFromEmail' already that determines the agency based on the email address. 

This would put the user's agency as the more specific office, which then we would use to filter on predictions. 

This seems a little contrary to what we discussed last week

File: server/routes/auth.routes.js [hunk 3]
```diff
@@ -232,32 +335,44 @@ function userOnPasswordOnlyWhitelist(session){
  * @return {*|PromiseLike<T | never>|Promise<T | never>}
  */
 function createMAXUser(cas_data) {
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:337]: Also we can remove the "createMaxUser" and updateMaxUser functions since we do not use MAX anymore.

Just make sure to verify the login as Dev button still works when you remove.

File: server/routes/prediction.routes.js [hunk 4]
```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
...
+    // Ensure user has necessary information
+    if (!user?.agency || !user?.userRole || !user?.email) {
+      logger.warn('Missing user info');
+      return { predictions: [], first: 0, rows: 0, totalCount: 0 };
     }
 
-    logger.debug("Entering getPredictions")
-    // await updatePredictionTable()
+    // Compute the user's office dynamically from their email
+    user.office = getOfficeFromEmail(user.email);
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:382]: Shouldn't the user.office already be populated after you login?

File: server/routes/prediction.routes.js [hunk 5]
```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
...
     }
 
-    // filter out rows
+    // Global text search
     if (filter.globalFilter) {
-      attributes.where.searchText = { [Op.like]: `%${filter.globalFilter.toLowerCase()}%` }
-    }
-    for (let f of ['office', 'agency', 'title', 'solNum', 'reviewRec', 'id']) {
-      normalizeMatchFilter(filter, f)
+      attributes.where.searchText = { [Sequelize.Op.like]: `%${filter.globalFilter.toLowerCase()}%` };
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:401]: Why do you have Sequelize.Op? Since we already imported "Op"

File: server/routes/prediction.routes.js [hunk 6]
```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
...
 
-
-    // process PrimeNG filters: filter.filters = { field: { value: 'x', matchMode: 'equals' } }
-    if (filter.filters) {
-      for (let f in filter.filters) {
-        if (filter.filters.hasOwnProperty(f) && filter.filters[f].matchMode === 'equals') {
-          attributes.where[f] = {[Op.eq]: filter.filters[f].value}
-        }
+    // Normalize filters
+    ['office', 'agency', 'title', 'solNum', 'reviewRec', 'id'].forEach(f => {
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:405]: I personally prefer separating out code into the small functions instead of having a large function with a lot of executing code.

File: server/routes/prediction.routes.js [hunk 7]
```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
...
+      Object.entries(filter.filters).forEach(([field, data]) => {
+        if (data.matchMode === 'equals' && data.value) {
+          attributes.where[field] = data.value; // Simplified without Op.eq
         }
+      });
+      logger.debug('Attributes after applying filters', { attributes });
+    }
+
+    // Handle date filtering
+    if (!filter.ignoreDateCutoff && !filter.filters?.solNum) {
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:425]: Probably can separate out into own function

File: server/routes/prediction.routes.js [hunk 8]
```diff
@@ -359,146 +364,155 @@ function normalizeMatchFilter(filter, field){
...
-    if ( ! authRoutes.isGSAAdmin(user.agency, user.userRole)) {
-      attributes.where.agency  = {
-        [Op.eq] : (user && user.agency) ? user.agency : ''
-      }
-    }
+    // Agency/Office filtering for non-admin users
+    if (!isGSAAdmin(user.agency, user.userRole)) {
+      // Apply agency and computed office filters directly
+      attributes.where.agency = user.agency;
+      attributes.where.office = user.office;
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:457]: How are you going to pull RFQ's that have a agency value that would normally be an office value?

For instance, Department of the Navy

File: server/routes/prediction.routes.js [hunk 9]
```diff
@@ -554,7 +568,7 @@ module.exports = {
     let keys = Object.keys(req.body)
 
     // verify that only supported filter params are used
-    let validKeys = ['agency', 'office', 'numDocs', 'solNum', 'category_list', 'startDate', 'fromPeriod', 'endDate', 'toPeriod', 'noticeType']
+    let validKeys = ['agency', 'office', 'numDocs', 'solNum', 'category_list', 'startDate', 'fromPeriod', 'endDate', 'toPeriod', 'userEmail']
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:571]: Please replace the userEmail back with the notice type. We are not going to be filtering Solicitations based on user emails.

File: server/routes/prediction.routes.js [hunk 10]
```diff
@@ -577,6 +591,9 @@ module.exports = {
     req.body.rows = (req.body.rows !== undefined) ? req.body.rows : 100
     let user = authRoutes.userInfoFromReq(req)
 
+     // Add `userEmail` from request body to the filter
+     req.body.userEmail = req.body.userEmail || user.email;
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:595]: You shouldn't need the user email for filtering if the agency & office are already provided

**Review by BuckinghamAJ — CHANGES_REQUESTED** [review]
# Things To Address
- [ ]  The SQL Migration file if we are moving to having a user office value, even though I thought we weren't. 
- [ ]  The filtering of Solicitations. 
- [ ] Code Comments

File: server/routes/auth.routes.js [hunk 11]
```diff
@@ -40,13 +40,23 @@ const FIVE08_COORDINATOR_ROLE = 2
...
+  'Department of Commerce': ['National Oceanic and Atmospheric Administration', 'National Telecommunications and Information Administration'],
+  'Department of the Interior': ['National Park Service', 'Fish and Wildlife Service', 'Bureau of Ocean Energy Management'],
+  'Department of the Treasury': ['Internal Revenue Service', 'U.S. Mint']
+};
+
 // Load your RSA private key
 let privateKey;
 try {
-  privateKey = fs.readFileSync(path.resolve(__dirname,'../certs/private.pem'), 'utf8');
+  privateKey = fs.readFileSync(path.resolve(__dirname, '../server/certs/private.pem'), 'utf8');
```
[the remark this answers is not in the document]
Comment (collinschreyer-dev) [reply server/routes/auth.routes.js:55]: Thanks for catching this!

File: server/routes/auth.routes.js [hunk 12]
```diff
@@ -57,16 +67,19 @@ try {
...
+      office: getOfficeFromEmail(cas_data['email-address']),
+      tag: 'updateMAXUser'
+    })
+
     user['firstName'] = cas_data['first-name']
     user['lastName'] = cas_data['last-name']
     user['email'] = cas_data['email-address']
     user['password'] = null
     user['agency'] = cas_data['agency-name']
+    user['office'] = getOfficeFromEmail(cas_data['email-address'])
```
[the remark this answers is not in the document]
Comment (collinschreyer-dev) [reply server/routes/auth.routes.js:82]: You’re correct; this was left over from my first attempt at filtering a few weeks back. I overlooked removing it. I went ahead and cleaned this up to align with what we agreed on. Thanks for catching that!

File: server/routes/auth.routes.js (hunk 1 above)
Comment (collinschreyer-dev) [reply server/routes/auth.routes.js:129]: You're right - I mistakenly removed something that's used later... I added back the user_email variable before creating the user_data object

File: server/routes/prediction.routes.js (hunk 4 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:382]: # Understanding Office Computation in SRT

## Current Flow

### Login (auth.routes.js)
```javascript
// In casStage2:
const office = getOfficeFromEmail(email);
req.session.cas_userinfo.office = office;

// In token creation (convertCASNamesToSRT):
const office = getOfficeFromEmail(srt_userinfo['email'])
srt_userinfo['office'] = office
```

### Predictions (prediction.routes.js)
```javascript
// In getPredictions:
user.office = getOfficeFromEmail(user.email);
```

## Why We Compute Office Again

1. Office is not stored in database
   - We decided to compute it dynamically rather than store it
   - User model doesn't have an office field

2. Token/Session contains office but:
   - We're computing it fresh each time for predictions
   - This ensures we always have the most current mapping

## Current Architecture
- Compute office from email when needed
- Don't persist in database
- Include in session/token info
- Recompute for predictions

So the line, `user.office = getOfficeFromEmail(user.email)` in getPredictions is necessary because we're following the pattern of computing office on demand rather than storing it permanently.

Alternatives could be:
1. Store office in database (decided against this)
2. Extract from token instead of recomputing
3. Keep current approach of computing when needed

File: server/routes/prediction.routes.js (hunk 6 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:405]: Thank you for this comment, I fell back to how I used to do things by making a giant, messy function.

The code has been completely refactored into smaller, focused functions:
* `applyBasicFilters` for notice types and global search
* `normalizeStandardFilters` for filter normalization
* `applyRegularFilters` for standard field filters
* `applyDateFilters` for date filtering
* `applyAgencyOfficeFilters` for agency/office logic
* `applySorting` for sort ordering

Each function has a single responsibility, making the code more maintainable and testable.

File: server/routes/prediction.routes.js (hunk 8 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:457]: The `applyAgencyOfficeFilters` function now handles this case by checking both fields:
```javascript
if (officeFilter) {
  attributes.where[Op.or] = [
    { office: officeFilter },
    { agency: officeFilter }  // This catches cases where EBUY data puts office in agency field
  ];
}
```

File: server/routes/prediction.routes.js (hunk 9 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:571]: The code completely removes any reference to userEmail filtering. Instead, it uses the notice type filter in applyBasicFilters

File: server/routes/prediction.routes.js (hunk 10 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:595]: All user email filtering has been removed.

File: server/routes/auth.routes.js (hunk 2 above)
Comment (BuckinghamAJ) [reply server/routes/auth.routes.js:244]: If you look grabAgencyFromEmail it utilizes a function called translateCASAgencyName, which grabs the value from Agency_Lookup in the config.js. 

So for instance if we add "army" to be Department of the Army I think that makes sense, and then start adding more values that we see are missing based on who is signing into SRT.

We can also look to add that information of office to agency in the config.js as well instead of in the auth route.

File: server/routes/prediction.routes.js [hunk 13]
```diff
@@ -366,131 +366,149 @@ function normalizeMatchFilter(filter, field){
...
+    
+    if (officeFilter) {
+      // Look for matches in both office and agency fields for the office value
+      attributes.where[Op.or] = [
+        { office: officeFilter },
+        { agency: officeFilter }  // This catches cases where EBUY data puts office in agency field
+      ];
+    } else {
+      // Default behavior - filter by user's office
       attributes.where.office = user.office;
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:448]: You are still filtering by user.office even though that will not exist

File: server/routes/auth.routes.js [hunk 14]
```diff
@@ -136,22 +132,19 @@ function getGovernmentEmail(emails) {
...
   let date = (now.getMonth() + 1) + "-" + now.getDate() + "-" + now.getFullYear();
-  let office = getOfficeFromEmail(loginGovUser.email);
-
+  const user_email = gov_email || loginGovUser.email
   let user_data = {
     'firstName': loginGovUser.given_name || null,
     'lastName': loginGovUser.family_name || null,
     'email': user_email,
     'password': null,
     'agency': grabAgencyFromEmail(user_email),
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:146]: Please add all the Agency email you found into the AGENCY_LOOKUP in the config.

For instance:
```
"army" : "Department of the Army",
"navy": "Department of the Navy",
etc.
```

File: server/routes/auth.routes.js [hunk 15]
```diff
@@ -119,36 +130,121 @@ function getGovernmentEmail(emails) {
...
+    .then((u) => {
+      return u;
     })
+    .catch((e) => {
+      logger.log("error", 'Error in: createUser', {error: e, tag: "createUser"});
+    });
+}
+
+
+const agencyNameVariations = {
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:166]: This can be in the config.js file. We should try our best to not clog up the code with large constants especially if we have a config file that we can use.

Then this could be ```agencyNameVariations = getConfig("<name in config>") ```

File: server/routes/auth.routes.js [hunk 16]
```diff
@@ -119,36 +130,121 @@ function getGovernmentEmail(emails) {
...
+    'Bureau of Ocean Energy Management': ['BOEM']
+  },
+  'Department of the Treasury': {
+    'Internal Revenue Service': ['IRS'],
+    'U.S. Mint': ['MINT', 'US MINT']
+  }
+};
+
+
+function getOfficeFromEmail(email) {
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:201]: All this info can be added to Agency_Lookup config file

File: server/routes/auth.routes.js (hunk 3 above)
Comment (BuckinghamAJ) [reply server/routes/auth.routes.js:330]: Just going to re-ping this since MAX.gov is not used anymore. Please let me know if we remove this createMaxUser and updateMaxUser if that will cause issues with the "Login as Dev" button

**Review by BuckinghamAJ — CHANGES_REQUESTED** [review]
Still some this to change.

- Adding info to config.js
- Instance of user.office in predictions.routes.js

File: server/routes/prediction.routes.js (hunk 13 above)
Comment (collinschreyer-dev) [reply server/routes/prediction.routes.js:448]: Modified applyAgencyOfficeFilters and a few other places to remove any type of querying to a user's office(non-existant)

File: server/routes/auth.routes.js (hunk 14 above)
Comment (collinschreyer-dev) [reply server/routes/auth.routes.js:146]: removed all these constants from the file

File: server/routes/auth.routes.js (hunk 15 above)
Comment (collinschreyer-dev) [reply server/routes/auth.routes.js:166]: # Refactoring Explanation: Using the Config File to Reduce Code Clutter

Here’s a detailed explanation of what we did to refactor the code in server/routes/auth.routes.js to use the configuration file instead of hardcoding constants in the routes file.

Before Refactoring

	1.	Hardcoded Constants in the Routes File:
	•	The agencyNameVariations object and getOfficeFromEmail logic were directly defined in auth.routes.js.
	•	This made the file cluttered and harder to maintain.

```
const agencyNameVariations = {
  "army": "Department of the Army",
  "navy": "Department of the Navy",
  // More entries here...
};
```

```
function getOfficeFromEmail(email) {
  if (email.includes("army.mil")) return "Department of the Army";
  if (email.includes("navy.mil")) return "Department of the Navy";
  // More logic here...
}
```


	2.	Constants and Logic Spread Across Code:
	•	Functions such as grabAgencyFromEmail and getOfficeFromEmail relied on hardcoded logic to map email domains to agencies and offices.
	•	Changes required editing multiple places in the code.
	3.	Duplicate Effort:
	•	The same variations and mappings (e.g., agencyNameVariations) could appear in other files, leading to duplication.

After Refactoring

1. Centralized Configuration:

	•	Added a structured AGENCY_HIERARCHY object to the configuration file.
	•	This object contains:
	•	Agencies and their offices.
	•	Aliases for each office.
	•	Email domains associated with each agency and office.

Config File Example:

```
AGENCY_HIERARCHY: {
  'Department of Defense': {
    offices: ['DEPARTMENT OF THE ARMY', 'DEPARTMENT OF THE NAVY'],
    variations: {
      'DEPARTMENT OF THE ARMY': {
        aliases: ['DEPT OF THE ARMY', 'US ARMY', 'ARMY'],
        email_domains: ['army.mil']
      },
      'DEPARTMENT OF THE NAVY': {
        aliases: ['DEPT OF THE NAVY', 'US NAVY', 'NAVY'],
        email_domains: ['navy.mil', 'us.navy.mil']
      }
    }
  }
}
```

2. Refactored auth.routes.js:

	•	Removed hardcoded constants like agencyNameVariations.
	•	Replaced getOfficeFromEmail with a function that references the AGENCY_HIERARCHY from the config.

Updated Logic in auth.routes.js:

```
const { getConfig } = require('../config'); // Assuming a helper function for accessing config
const hierarchy = getConfig('AGENCY_HIERARCHY');

// Function to derive office from email
function getOfficeFromEmail(email) {
  if (!email) return null;

  const emailLower = email.toLowerCase();
  for (const [agency, agencyData] of Object.entries(hierarchy)) {
    for (const [officeName, officeData] of Object.entries(agencyData.variations)) {
      if (officeData.email_domains.some(domain => emailLower.includes(domain))) {
        return officeName;
      }
    }
  }
  return null;
}
```

```
// Updated user creation
let office = getOfficeFromEmail(loginGovUser.email);
let user_data = {
  firstName: loginGovUser.given_name || null,
  lastName: loginGovUser.family_name || null,
  email: gov_email || loginGovUser.email,
  agency: grabAgencyFromEmail(gov_email || loginGovUser.email),
  office: office
};
```

3. Dynamic Lookup:

	•	getOfficeFromEmail now dynamically retrieves mappings from the AGENCY_HIERARCHY instead of relying on hardcoded logic.



Summary

	•	Moved hardcoded agency and office mappings to AGENCY_HIERARCHY in config.js.
	•	Updated auth.routes.js to dynamically retrieve agency and office mappings using the configuration.
	•	Reduced duplication and improved maintainability by centralizing all agency-related data in the config file.

File: server/config/config.js
Comment (BuckinghamAJ) [inline server/config/config.js:1]: Can you please add what you had in the predictions route into the AGENCY_LOOKUP

File: server/routes/auth.routes.js [hunk 17]
```diff
@@ -662,41 +705,56 @@ module.exports = {
    * @param req
    * @param res
    */
-  casStage2 : async function (req, res) {
-    /** @namespace req.session.cas_userinfo */
-
-    if ( ! ( req.session && req.session['cas_userinfo'] && (req.session.cas_userinfo['max-id'] || req.session.cas_userinfo['maxId']  ))) {
-      // didn't get CAS session info
+  casStage2: async function(req, res) {
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js:708]: Not sure why you updated this since, this only affects Max.gov login which is not being utilized anymore. Was there a reason?

File: server/routes/auth.routes.js [hunk 18]
```diff
@@ -162,89 +154,30 @@ function createUser(loginGovUser) {
...
- 
-  // Interior Mappings
-  if (emailLower.includes('nps.gov')) return 'NATIONAL PARK SERVICE';
-  if (emailLower.includes('fws.gov')) return 'US FISH AND WILDLIFE SERVICE';
-  if (emailLower.includes('boem.gov')) return 'BUREAU OF OCEAN ENERGY MANAGEMENT';
- 
-  // Treasury Mappings
-  if (emailLower.includes('irs.gov')) return 'INTERNAL REVENUE SERVICE';
-  if (emailLower.includes('usmint.treas.gov')) return 'US MINT';
-
```
Comment (BuckinghamAJ) [inline server/routes/auth.routes.js]: I would like all the stuff above added to the config.js under the AGENCY_LOOKUP if the acronyms are not already there.

File: server/routes/prediction.routes.js [hunk 19]
```diff
@@ -434,31 +452,59 @@ const applyDateFilters = (attributes, filter, config) => {
...
-        { office: officeFilter },
-        { agency: officeFilter }  // This catches cases where EBUY data puts office in agency field
-      ];
-    } else {
-      // Default behavior - filter by user's office
-      attributes.where.office = user.office;
-    }
+    const agencyFilter = filter.filters?.agency?.value;
+
+    if (!agencyFilter && !officeFilter && user.office) {
```
Comment (BuckinghamAJ) [inline server/routes/prediction.routes.js:462]: Very confused as to why we are still utilizing user.office.

**Review by BuckinghamAJ — CHANGES_REQUESTED** [review]
